Add some Cocoa helper templates for getting objects of types and filtering containers.

xeenon opened the pull request:
#### 485052b30901b36de44f76c560e60674c1518c76
<pre>
Add some Cocoa helper templates for getting objects of types and filtering containers.
<a href="https://bugs.webkit.org/show_bug.cgi?id=245646">https://bugs.webkit.org/show_bug.cgi?id=245646</a>

Reviewed by Tim Horton.

These will be used in some forthcoming feature work. Instead of adding them as category methods in WebKit,
I made them template functions to avoid cluttering up Foundation classes with WebKit category methods.

* Source/WebKit/Platform/cocoa/CocoaHelpers.h: Added.
(WebKit::filterObjects):
(WebKit::mapObjects):
(WebKit::objectForKey):
* Source/WebKit/Platform/cocoa/CocoaHelpers.mm: Added.
(WebKit::filterObjects&lt;NSArray&gt;):
(WebKit::filterObjects&lt;NSDictionary&gt;):
(WebKit::filterObjects&lt;NSSet&gt;):
(WebKit::mapObjects&lt;NSArray&gt;):
(WebKit::mapObjects&lt;NSDictionary&gt;):
(WebKit::mapObjects&lt;NSSet&gt;):
(WebKit::objectForKey&lt;NSString&gt;):
(WebKit::objectForKey&lt;NSArray&gt;):
(WebKit::objectForKey&lt;NSDictionary&gt;):
(WebKit::objectForKey&lt;NSSet&gt;):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/254877@main">https://commits.webkit.org/254877@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f97c9386c9500aef456481c0e73b6ecb4f639c5a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/90562 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/35146 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/21200 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/8/builds/99894 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/33646 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/28831 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/82927 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/96329 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/96217 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/68/builds/26779 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/77407 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/82927 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/81557 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/21200 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/82927 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/34745 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/21200 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/32560 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/21200 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/74/builds/3407 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/36325 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/77407 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/1476 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/38239 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/21200 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
```diff
@@ -0,0 +1,189 @@
+template<>
+NSDictionary *filterObjects<NSDictionary>(NSDictionary *dictionay, bool NS_NOESCAPE (^block)(__kindof id key, __kindof id value))
+{
+    if (!dictionay.count)
```
sp. "dictionay"

hortont424 commented on Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
```diff
@@ -0,0 +1,189 @@
+template<>
+NSDictionary *mapObjects<NSDictionary>(NSDictionary *dictionay, __kindof id NS_NOESCAPE (^block)(__kindof id key, __kindof id value))
+{
+    if (!dictionay.count)
```
and again

hortont424 reviewed (approved):
I wonder if "Cocoa" is the right word or we want "Foundation".

xeenon committed:
Add some Cocoa helper templates for getting objects of types and filtering containers.
https://bugs.webkit.org/show_bug.cgi?id=245646

Reviewed by Tim Horton.

These will be used in some forthcoming feature work. Instead of adding them as category methods in WebKit,
I made them template functions to avoid cluttering up Foundation classes with WebKit category methods.

* Source/WebKit/Platform/cocoa/CocoaHelpers.h: Added.
(WebKit::filterObjects):
(WebKit::mapObjects):
(WebKit::objectForKey):
* Source/WebKit/Platform/cocoa/CocoaHelpers.mm: Added.
(WebKit::filterObjects<NSArray>):
(WebKit::filterObjects<NSDictionary>):
(WebKit::filterObjects<NSSet>):
(WebKit::mapObjects<NSArray>):
(WebKit::mapObjects<NSDictionary>):
(WebKit::mapObjects<NSSet>):
(WebKit::objectForKey<NSString>):
(WebKit::objectForKey<NSArray>):
(WebKit::objectForKey<NSDictionary>):
(WebKit::objectForKey<NSSet>):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/254877@main
